opentelemetry/src/baggage.rs

```diff
@@ -282,10 +281,53 @@ impl FromIterator<KeyValueMetadata> for Baggage {
     }
 }
 
+fn encode(s: &str) -> String {
+    let special_characters = [b'.', b'-', b'_', b'~'];
+    let mut encoded_string = String::with_capacity(s.len() * 3);
+
+    let bytes = s.as_bytes();
+    let mut i = 0;
+
+    while i < bytes.len() {
+        let byte = bytes[i];
+
+        match byte {
+            b' ' => encoded_string.push_str("%20"),
+            byte if byte.is_ascii_alphanumeric() || special_characters.contains(&byte) => {
+                encoded_string.push(byte as char)
+            }
+            _ => {
+                if byte.is_ascii() {
+                    encoded_string.push_str(&format!("%{:02X}", byte));
+                } else {
+                    let start = i;
+                    let mut end = start + 1;
+
+                    while end < bytes.len() && !bytes[end].is_ascii() {
+                        end += 1;
+                    }
+
+                    // Encoding each byte of the multi-byte character
+                    for &multi_byte in &bytes[start..end] {
+                        encoded_string.push_str(&format!("%{:02X}", multi_byte));
+                    }
+
+                    // Adjust `i` to skip over the bytes we've just encoded
+                    i = end - 1;
+                }
+            }
+        }
+
+        i += 1;
+    }
+
+    encoded_string
+}
```

Comment:
@sreeo Not sure if I am missing something, but we don't need separate handling for ASCII (non-alphanumeric) and UTF-8. Both need to be byte-wise percentile-encoded. I tested your unit-test examples on something I suggested earlier, and seems to work - https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1f433e6f80889b0680bab004f926ecaa.  Though it is difficult to estimate the encoded size, it should be fine to have initially set to the original string size, and not over-estimate. Sorry for being bit critical from perf prospective here :)

Reply:
No worries, your suggested code is simpler than the current one and passes all tests. Replaced it.